// @annotorious/react/dist/annotorious-react.es17.js
import { computePosition as I, inline as J, offset as K, flip as N, shift as Q, arrow as L } from "./annotorious-react.es16.js";
import { autoUpdate as ce, platform as ae } from "./annotorious-react.es16.js";
import * as f from "react";
import { useLayoutEffect as T } from "react";
import * as X from "react-dom";
var Y = typeof document < "u", Z = function() {
}, R = Y ? T : Z;
function h(e, n) {
  if (e === n)
    return !0;
  if (typeof e != typeof n)
    return !1;
  if (typeof e == "function" && e.toString() === n.toString())
    return !0;
  let t, r, s;
  if (e && n && typeof e == "object") {
    if (Array.isArray(e)) {
      if (t = e.length, t !== n.length) return !1;
      for (r = t; r-- !== 0; )
        if (!h(e[r], n[r]))
          return !1;
      return !0;
    }
    if (s = Object.keys(e), t = s.length, t !== Object.keys(n).length)
      return !1;
    for (r = t; r-- !== 0; )
      if (!{}.hasOwnProperty.call(n, s[r]))
        return !1;
    for (r = t; r-- !== 0; ) {
      const p = s[r];
      if (!(p === "_owner" && e.$$typeof) && !h(e[p], n[p]))
        return !1;
    }
    return !0;
  }
  return e !== e && n !== n;
}
function q(e) {
  return typeof window > "u" ? 1 : (e.ownerDocument.defaultView || window).devicePixelRatio || 1;
}
function A(e, n) {
  const t = q(e);
  return Math.round(n * t) / t;
}
function k(e) {
  const n = f.useRef(e);
  return R(() => {
    n.current = e;
  }), n;
}
function ne(e) {
  e === void 0 && (e = {});
  const {
    placement: n = "bottom",
    strategy: t = "absolute",
    middleware: r = [],
    platform: s,
    elements: {
      reference: p,
      floating: B
    } = {},
    transform: D = !0,
    whileElementsMounted: E,
    open: x
  } = e, [u, C] = f.useState({
    x: 0,
    y: 0,
    strategy: t,
    placement: n,
    middlewareData: {},
    isPositioned: !1
  }), [M, U] = f.useState(r);
  h(M, r) || U(r);
  const [V, W] = f.useState(null), [z, G] = f.useState(null), O = f.useCallback((o) => {
    o !== a.current && (a.current = o, W(o));
  }, []), _ = f.useCallback((o) => {
    o !== m.current && (m.current = o, G(o));
  }, []), i = p || V, l = B || z, a = f.useRef(null), m = f.useRef(null), y = f.useRef(u), H = E != null, P = k(E), S = k(s), v = k(x), g = f.useCallback(() => {
    if (!a.current || !m.current)
      return;
    const o = {
      placement: n,
      strategy: t,
      middleware: M
    };
    S.current && (o.platform = S.current), I(a.current, m.current, o).then((w) => {
      const d = {
        ...w,
        // The floating element's position may be recomputed while it's closed
        // but still mounted (such as when transitioning out). To ensure
        // `isPositioned` will be `false` initially on the next open, avoid
        // setting it to `true` when `open === false` (must be specified).
        isPositioned: v.current !== !1
      };
      $.current && !h(y.current, d) && (y.current = d, X.flushSync(() => {
        C(d);
      }));
    });
  }, [M, n, t, S, v]);
  R(() => {
    x === !1 && y.current.isPositioned && (y.current.isPositioned = !1, C((o) => ({
      ...o,
      isPositioned: !1
    })));
  }, [x]);
  const $ = f.useRef(!1);
  R(() => ($.current = !0, () => {
    $.current = !1;
  }), []), R(() => {
    if (i && (a.current = i), l && (m.current = l), i && l) {
      if (P.current)
        return P.current(i, l, g);
      g();
    }
  }, [i, l, g, P, H]);
  const F = f.useMemo(() => ({
    reference: a,
    floating: m,
    setReference: O,
    setFloating: _
  }), [O, _]), c = f.useMemo(() => ({
    reference: i,
    floating: l
  }), [i, l]), j = f.useMemo(() => {
    const o = {
      position: t,
      left: 0,
      top: 0
    };
    if (!c.floating)
      return o;
    const w = A(c.floating, u.x), d = A(c.floating, u.y);
    return D ? {
      ...o,
      transform: "translate(" + w + "px, " + d + "px)",
      ...q(c.floating) >= 1.5 && {
        willChange: "transform"
      }
    } : {
      position: t,
      left: w,
      top: d
    };
  }, [t, D, c.floating, u.x, u.y]);
  return f.useMemo(() => ({
    ...u,
    update: g,
    refs: F,
    elements: c,
    floatingStyles: j
  }), [u, g, F, c, j]);
}
const b = (e) => {
  function n(t) {
    return {}.hasOwnProperty.call(t, "current");
  }
  return {
    name: "arrow",
    options: e,
    fn(t) {
      const {
        element: r,
        padding: s
      } = typeof e == "function" ? e(t) : e;
      return r && n(r) ? r.current != null ? L({
        element: r.current,
        padding: s
      }).fn(t) : {} : r ? L({
        element: r,
        padding: s
      }).fn(t) : {};
    }
  };
}, re = (e, n) => {
  const t = K(e);
  return {
    name: t.name,
    fn: t.fn,
    options: [e, n]
  };
}, fe = (e, n) => {
  const t = Q(e);
  return {
    name: t.name,
    fn: t.fn,
    options: [e, n]
  };
}, oe = (e, n) => {
  const t = N(e);
  return {
    name: t.name,
    fn: t.fn,
    options: [e, n]
  };
}, se = (e, n) => {
  const t = J(e);
  return {
    name: t.name,
    fn: t.fn,
    options: [e, n]
  };
}, ue = (e, n) => {
  const t = b(e);
  return {
    name: t.name,
    fn: t.fn,
    options: [e, n]
  };
};
export {
  ue as arrow,
  ce as autoUpdate,
  I as computePosition,
  oe as flip,
  se as inline,
  re as offset,
  ae as platform,
  fe as shift,
  ne as useFloating
};
