// @annotorious/react/dist/annotorious-react.es16.js
import { computePosition as st, inline as rt, offset as ct, flip as lt, shift as ft, arrow as ut } from "./annotorious-react.es24.js";
import { createCoords as m, rectToClientRect as J, round as A, max as E, min as V, floor as D } from "./annotorious-react.es23.js";
import { getOverflowAncestors as z, isElement as v, getDocumentElement as b, getWindow as O, getFrameElement as U, getComputedStyle as R, isHTMLElement as T, isTopLayer as $, getParentNode as W, isLastTraversableNode as H, isTableElement as at, isContainingBlock as Q, getContainingBlock as ht, getNodeName as _, isOverflowElement as q, getNodeScroll as I, isWebKit as Y } from "./annotorious-react.es20.js";
function Z(t) {
  const e = R(t);
  let n = parseFloat(e.width) || 0, o = parseFloat(e.height) || 0;
  const i = T(t), r = i ? t.offsetWidth : n, s = i ? t.offsetHeight : o, c = A(n) !== r || A(o) !== s;
  return c && (n = r, o = s), {
    width: n,
    height: o,
    $: c
  };
}
function X(t) {
  return v(t) ? t : t.contextElement;
}
function F(t) {
  const e = X(t);
  if (!T(e))
    return m(1);
  const n = e.getBoundingClientRect(), {
    width: o,
    height: i,
    $: r
  } = Z(e);
  let s = (r ? A(n.width) : n.width) / o, c = (r ? A(n.height) : n.height) / i;
  return (!s || !Number.isFinite(s)) && (s = 1), (!c || !Number.isFinite(c)) && (c = 1), {
    x: s,
    y: c
  };
}
const dt = /* @__PURE__ */ m(0);
function tt(t) {
  const e = O(t);
  return !Y() || !e.visualViewport ? dt : {
    x: e.visualViewport.offsetLeft,
    y: e.visualViewport.offsetTop
  };
}
function gt(t, e, n) {
  return e === void 0 && (e = !1), !n || e && n !== O(t) ? !1 : e;
}
function L(t, e, n, o) {
  e === void 0 && (e = !1), n === void 0 && (n = !1);
  const i = t.getBoundingClientRect(), r = X(t);
  let s = m(1);
  e && (o ? v(o) && (s = F(o)) : s = F(t));
  const c = gt(r, n, o) ? tt(r) : m(0);
  let l = (i.left + c.x) / s.x, f = (i.top + c.y) / s.y, u = i.width / s.x, h = i.height / s.y;
  if (r) {
    const d = O(r), a = o && v(o) ? O(o) : o;
    let w = d, p = U(w);
    for (; p && o && a !== w; ) {
      const y = F(p), g = p.getBoundingClientRect(), x = R(p), C = g.left + (p.clientLeft + parseFloat(x.paddingLeft)) * y.x, S = g.top + (p.clientTop + parseFloat(x.paddingTop)) * y.y;
      l *= y.x, f *= y.y, u *= y.x, h *= y.y, l += C, f += S, w = O(p), p = U(w);
    }
  }
  return J({
    width: u,
    height: h,
    x: l,
    y: f
  });
}
function M(t, e) {
  const n = I(t).scrollLeft;
  return e ? e.left + n : L(b(t)).left + n;
}
function et(t, e) {
  const n = t.getBoundingClientRect(), o = n.left + e.scrollLeft - M(t, n), i = n.top + e.scrollTop;
  return {
    x: o,
    y: i
  };
}
function pt(t) {
  let {
    elements: e,
    rect: n,
    offsetParent: o,
    strategy: i
  } = t;
  const r = i === "fixed", s = b(o), c = e ? $(e.floating) : !1;
  if (o === s || c && r)
    return n;
  let l = {
    scrollLeft: 0,
    scrollTop: 0
  }, f = m(1);
  const u = m(0), h = T(o);
  if ((h || !h && !r) && ((_(o) !== "body" || q(s)) && (l = I(o)), h)) {
    const a = L(o);
    f = F(o), u.x = a.x + o.clientLeft, u.y = a.y + o.clientTop;
  }
  const d = s && !h && !r ? et(s, l) : m(0);
  return {
    width: n.width * f.x,
    height: n.height * f.y,
    x: n.x * f.x - l.scrollLeft * f.x + u.x + d.x,
    y: n.y * f.y - l.scrollTop * f.y + u.y + d.y
  };
}
function wt(t) {
  return Array.from(t.getClientRects());
}
function yt(t) {
  const e = b(t), n = I(t), o = t.ownerDocument.body, i = E(e.scrollWidth, e.clientWidth, o.scrollWidth, o.clientWidth), r = E(e.scrollHeight, e.clientHeight, o.scrollHeight, o.clientHeight);
  let s = -n.scrollLeft + M(t);
  const c = -n.scrollTop;
  return R(o).direction === "rtl" && (s += E(e.clientWidth, o.clientWidth) - i), {
    width: i,
    height: r,
    x: s,
    y: c
  };
}
const K = 25;
function mt(t, e) {
  const n = O(t), o = b(t), i = n.visualViewport;
  let r = o.clientWidth, s = o.clientHeight, c = 0, l = 0;
  if (i) {
    r = i.width, s = i.height;
    const u = Y();
    (!u || u && e === "fixed") && (c = i.offsetLeft, l = i.offsetTop);
  }
  const f = M(o);
  if (f <= 0) {
    const u = o.ownerDocument, h = u.body, d = getComputedStyle(h), a = u.compatMode === "CSS1Compat" && parseFloat(d.marginLeft) + parseFloat(d.marginRight) || 0, w = Math.abs(o.clientWidth - h.clientWidth - a);
    w <= K && (r -= w);
  } else f <= K && (r += f);
  return {
    width: r,
    height: s,
    x: c,
    y: l
  };
}
function xt(t, e) {
  const n = L(t, !0, e === "fixed"), o = n.top + t.clientTop, i = n.left + t.clientLeft, r = T(t) ? F(t) : m(1), s = t.clientWidth * r.x, c = t.clientHeight * r.y, l = i * r.x, f = o * r.y;
  return {
    width: s,
    height: c,
    x: l,
    y: f
  };
}
function j(t, e, n) {
  let o;
  if (e === "viewport")
    o = mt(t, n);
  else if (e === "document")
    o = yt(b(t));
  else if (v(e))
    o = xt(e, n);
  else {
    const i = tt(t);
    o = {
      x: e.x - i.x,
      y: e.y - i.y,
      width: e.width,
      height: e.height
    };
  }
  return J(o);
}
function nt(t, e) {
  const n = W(t);
  return n === e || !v(n) || H(n) ? !1 : R(n).position === "fixed" || nt(n, e);
}
function vt(t, e) {
  const n = e.get(t);
  if (n)
    return n;
  let o = z(t, [], !1).filter((c) => v(c) && _(c) !== "body"), i = null;
  const r = R(t).position === "fixed";
  let s = r ? W(t) : t;
  for (; v(s) && !H(s); ) {
    const c = R(s), l = Q(s);
    !l && c.position === "fixed" && (i = null), (r ? !l && !i : !l && c.position === "static" && !!i && (i.position === "absolute" || i.position === "fixed") || q(s) && !l && nt(t, s)) ? o = o.filter((u) => u !== s) : i = c, s = W(s);
  }
  return e.set(t, o), o;
}
function Rt(t) {
  let {
    element: e,
    boundary: n,
    rootBoundary: o,
    strategy: i
  } = t;
  const s = [...n === "clippingAncestors" ? $(e) ? [] : vt(e, this._c) : [].concat(n), o], c = j(e, s[0], i);
  let l = c.top, f = c.right, u = c.bottom, h = c.left;
  for (let d = 1; d < s.length; d++) {
    const a = j(e, s[d], i);
    l = E(a.top, l), f = V(a.right, f), u = V(a.bottom, u), h = E(a.left, h);
  }
  return {
    width: f - h,
    height: u - l,
    x: h,
    y: l
  };
}
function bt(t) {
  const {
    width: e,
    height: n
  } = Z(t);
  return {
    width: e,
    height: n
  };
}
function Ct(t, e, n) {
  const o = T(e), i = b(e), r = n === "fixed", s = L(t, !0, r, e);
  let c = {
    scrollLeft: 0,
    scrollTop: 0
  };
  const l = m(0);
  function f() {
    l.x = M(i);
  }
  if (o || !o && !r)
    if ((_(e) !== "body" || q(i)) && (c = I(e)), o) {
      const a = L(e, !0, r, e);
      l.x = a.x + e.clientLeft, l.y = a.y + e.clientTop;
    } else i && f();
  r && !o && i && f();
  const u = i && !o && !r ? et(i, c) : m(0), h = s.left + c.scrollLeft - l.x - u.x, d = s.top + c.scrollTop - l.y - u.y;
  return {
    x: h,
    y: d,
    width: s.width,
    height: s.height
  };
}
function N(t) {
  return R(t).position === "static";
}
function G(t, e) {
  if (!T(t) || R(t).position === "fixed")
    return null;
  if (e)
    return e(t);
  let n = t.offsetParent;
  return b(t) === n && (n = n.ownerDocument.body), n;
}
function ot(t, e) {
  const n = O(t);
  if ($(t))
    return n;
  if (!T(t)) {
    let i = W(t);
    for (; i && !H(i); ) {
      if (v(i) && !N(i))
        return i;
      i = W(i);
    }
    return n;
  }
  let o = G(t, e);
  for (; o && at(o) && N(o); )
    o = G(o, e);
  return o && H(o) && N(o) && !Q(o) ? n : o || ht(t) || n;
}
const Ot = async function(t) {
  const e = this.getOffsetParent || ot, n = this.getDimensions, o = await n(t.floating);
  return {
    reference: Ct(t.reference, await e(t.floating), t.strategy),
    floating: {
      x: 0,
      y: 0,
      width: o.width,
      height: o.height
    }
  };
};
function Lt(t) {
  return R(t).direction === "rtl";
}
const Tt = {
  convertOffsetParentRelativeRectToViewportRelativeRect: pt,
  getDocumentElement: b,
  getClippingRect: Rt,
  getOffsetParent: ot,
  getElementRects: Ot,
  getClientRects: wt,
  getDimensions: bt,
  getScale: F,
  isElement: v,
  isRTL: Lt
};
function it(t, e) {
  return t.x === e.x && t.y === e.y && t.width === e.width && t.height === e.height;
}
function Et(t, e) {
  let n = null, o;
  const i = b(t);
  function r() {
    var c;
    clearTimeout(o), (c = n) == null || c.disconnect(), n = null;
  }
  function s(c, l) {
    c === void 0 && (c = !1), l === void 0 && (l = 1), r();
    const f = t.getBoundingClientRect(), {
      left: u,
      top: h,
      width: d,
      height: a
    } = f;
    if (c || e(), !d || !a)
      return;
    const w = D(h), p = D(i.clientWidth - (u + d)), y = D(i.clientHeight - (h + a)), g = D(u), C = {
      rootMargin: -w + "px " + -p + "px " + -y + "px " + -g + "px",
      threshold: E(0, V(1, l)) || 1
    };
    let S = !0;
    function k(P) {
      const B = P[0].intersectionRatio;
      if (B !== l) {
        if (!S)
          return s();
        B ? s(!1, B) : o = setTimeout(() => {
          s(!1, 1e-7);
        }, 1e3);
      }
      B === 1 && !it(f, t.getBoundingClientRect()) && s(), S = !1;
    }
    try {
      n = new IntersectionObserver(k, {
        ...C,
        // Handle <iframe>s
        root: i.ownerDocument
      });
    } catch {
      n = new IntersectionObserver(k, C);
    }
    n.observe(t);
  }
  return s(!0), r;
}
function Bt(t, e, n, o) {
  o === void 0 && (o = {});
  const {
    ancestorScroll: i = !0,
    ancestorResize: r = !0,
    elementResize: s = typeof ResizeObserver == "function",
    layoutShift: c = typeof IntersectionObserver == "function",
    animationFrame: l = !1
  } = o, f = X(t), u = i || r ? [...f ? z(f) : [], ...e ? z(e) : []] : [];
  u.forEach((g) => {
    i && g.addEventListener("scroll", n, {
      passive: !0
    }), r && g.addEventListener("resize", n);
  });
  const h = f && c ? Et(f, n) : null;
  let d = -1, a = null;
  s && (a = new ResizeObserver((g) => {
    let [x] = g;
    x && x.target === f && a && e && (a.unobserve(e), cancelAnimationFrame(d), d = requestAnimationFrame(() => {
      var C;
      (C = a) == null || C.observe(e);
    })), n();
  }), f && !l && a.observe(f), e && a.observe(e));
  let w, p = l ? L(t) : null;
  l && y();
  function y() {
    const g = L(t);
    p && !it(p, g) && n(), p = g, w = requestAnimationFrame(y);
  }
  return n(), () => {
    var g;
    u.forEach((x) => {
      i && x.removeEventListener("scroll", n), r && x.removeEventListener("resize", n);
    }), h == null || h(), (g = a) == null || g.disconnect(), a = null, l && cancelAnimationFrame(w);
  };
}
const Dt = ct, At = ft, Ht = lt, It = ut, Mt = rt, Nt = (t, e, n) => {
  const o = /* @__PURE__ */ new Map(), i = {
    platform: Tt,
    ...n
  }, r = {
    ...i.platform,
    _c: o
  };
  return st(t, e, {
    ...i,
    platform: r
  });
};
export {
  It as arrow,
  Bt as autoUpdate,
  Nt as computePosition,
  Ht as flip,
  z as getOverflowAncestors,
  Mt as inline,
  Dt as offset,
  Tt as platform,
  At as shift
};
