// @annotorious/react/dist/annotorious-react.es23.js
const m = Math.min, h = Math.max, S = Math.round, w = Math.floor, C = (t) => ({
  x: t,
  y: t
}), p = {
  left: "right",
  right: "left",
  bottom: "top",
  top: "bottom"
};
function j(t, n, e) {
  return h(t, m(n, e));
}
function L(t, n) {
  return typeof t == "function" ? t(n) : t;
}
function g(t) {
  return t.split("-")[0];
}
function a(t) {
  return t.split("-")[1];
}
function x(t) {
  return t === "x" ? "y" : "x";
}
function b(t) {
  return t === "y" ? "height" : "width";
}
function d(t) {
  const n = t[0];
  return n === "t" || n === "b" ? "y" : "x";
}
function A(t) {
  return x(d(t));
}
function E(t, n, e) {
  e === void 0 && (e = !1);
  const r = a(t), i = A(t), o = b(i);
  let c = i === "x" ? r === (e ? "end" : "start") ? "right" : "left" : r === "start" ? "bottom" : "top";
  return n.reference[o] > n.floating[o] && (c = u(c)), [c, u(c)];
}
function R(t) {
  const n = u(t);
  return [s(t), n, s(n)];
}
function s(t) {
  return t.includes("start") ? t.replace("start", "end") : t.replace("end", "start");
}
const f = ["left", "right"], l = ["right", "left"], P = ["top", "bottom"], y = ["bottom", "top"];
function O(t, n, e) {
  switch (t) {
    case "top":
    case "bottom":
      return e ? n ? l : f : n ? f : l;
    case "left":
    case "right":
      return n ? P : y;
    default:
      return [];
  }
}
function T(t, n, e, r) {
  const i = a(t);
  let o = O(g(t), e === "start", r);
  return i && (o = o.map((c) => c + "-" + i), n && (o = o.concat(o.map(s)))), o;
}
function u(t) {
  const n = g(t);
  return p[n] + t.slice(n.length);
}
function M(t) {
  return {
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    ...t
  };
}
function k(t) {
  return typeof t != "number" ? M(t) : {
    top: t,
    right: t,
    bottom: t,
    left: t
  };
}
function q(t) {
  const {
    x: n,
    y: e,
    width: r,
    height: i
  } = t;
  return {
    width: r,
    height: i,
    top: e,
    left: n,
    right: n + r,
    bottom: e + i,
    x: n,
    y: e
  };
}
export {
  j as clamp,
  C as createCoords,
  L as evaluate,
  M as expandPaddingObject,
  w as floor,
  a as getAlignment,
  A as getAlignmentAxis,
  E as getAlignmentSides,
  b as getAxisLength,
  R as getExpandedPlacements,
  s as getOppositeAlignmentPlacement,
  x as getOppositeAxis,
  T as getOppositeAxisPlacements,
  u as getOppositePlacement,
  k as getPaddingObject,
  g as getSide,
  d as getSideAxis,
  h as max,
  m as min,
  q as rectToClientRect,
  S as round
};
