// @annotorious/core/dist/annotorious-core.es.js
var G = Object.prototype.hasOwnProperty;
function $(e, t) {
  var n, o;
  if (e === t) return !0;
  if (e && t && (n = e.constructor) === t.constructor) {
    if (n === Date) return e.getTime() === t.getTime();
    if (n === RegExp) return e.toString() === t.toString();
    if (n === Array) {
      if ((o = e.length) === t.length)
        for (; o-- && $(e[o], t[o]); ) ;
      return o === -1;
    }
    if (!n || typeof e == "object") {
      o = 0;
      for (n in e)
        if (G.call(e, n) && ++o && !G.call(t, n) || !(n in t) || !$(e[n], t[n])) return !1;
      return Object.keys(t).length === o;
    }
  }
  return e !== e && t !== t;
}
const X = Symbol("clean");
let N = [], _ = 0;
const z = 4, Z = globalThis.nanostoresGlobal || (globalThis.nanostoresGlobal = { epoch: 0 }), M = /* @__NO_SIDE_EFFECTS__ */ (e) => {
  let t = [], n = {
    get() {
      return n.lc || n.listen(() => {
      })(), n.value;
    },
    init: e,
    lc: 0,
    listen(o) {
      return n.lc = t.push(o), () => {
        for (let f = _ + z; f < N.length; )
          N[f] === o ? N.splice(f, z) : f += z;
        let s = t.indexOf(o);
        ~s && (t.splice(s, 1), --n.lc || n.off());
      };
    },
    notify(o, s) {
      Z.epoch++;
      let f = !N.length;
      for (let A of t)
        N.push(A, n.value, o, s);
      if (f) {
        for (_ = 0; _ < N.length; _ += z)
          N[_](
            N[_ + 1],
            N[_ + 2],
            N[_ + 3]
          );
        N.length = 0;
      }
    },
    /* It will be called on last listener unsubscribing.
       We will redefine it in onMount and onStop. */
    off() {
    },
    set(o) {
      let s = n.value;
      s !== o && (n.value = o, n.notify(s));
    },
    subscribe(o) {
      let s = n.listen(o);
      return o(n.value), s;
    },
    value: e
  };
  return process.env.NODE_ENV !== "production" && (n[X] = () => {
    t = [], n.lc = 0, n.off();
  }), n;
}, Oe = (e) => {
  const t = /* @__PURE__ */ M(null);
  return e.observe(({ changes: n }) => {
    const o = t.get();
    if (o) {
      (n.deleted || []).some((A) => A.id === o) && t.set(null);
      const f = (n.updated || []).find(({ oldValue: A }) => A.id === o);
      f && t.set(f.newValue.id);
    }
  }), {
    get current() {
      return t.get();
    },
    subscribe: t.subscribe.bind(t),
    set: t.set.bind(t)
  };
};
var F = /* @__PURE__ */ ((e) => (e.EDIT = "EDIT", e.SELECT = "SELECT", e.NONE = "NONE", e))(F || {});
const P = { selected: [] }, Be = (e, t, n) => {
  const o = /* @__PURE__ */ M(P);
  let s = t;
  const f = () => {
    $(o.get(), P) || o.set(P);
  }, A = () => {
    var g;
    return ((g = o.get().selected) == null ? void 0 : g.length) === 0;
  }, w = (g) => {
    if (A())
      return !1;
    const b = typeof g == "string" ? g : g.id;
    return o.get().selected.some((L) => L.id === b);
  }, C = (g, b) => {
    let L;
    if (Array.isArray(g)) {
      if (L = g.map((i) => e.getAnnotation(i)).filter(Boolean), L.length < g.length) {
        console.warn("Invalid selection: " + g.filter((i) => !L.some((u) => u.id === i)));
        return;
      }
    } else {
      const i = e.getAnnotation(g);
      if (!i) {
        console.warn("Invalid selection: " + g);
        return;
      }
      L = [i];
    }
    const B = L.reduce((i, u) => {
      const h = v(u);
      return h === "EDIT" ? [...i, { id: u.id, editable: !0 }] : h === "SELECT" ? [...i, { id: u.id }] : i;
    }, []);
    o.set({ selected: B, event: b });
  }, m = (g, b) => {
    const L = Array.isArray(g) ? g : [g], B = L.map((i) => e.getAnnotation(i)).filter((i) => !!i);
    o.set({
      selected: B.map((i) => {
        const u = b === void 0 ? v(i) === "EDIT" : b;
        return { id: i.id, editable: u };
      })
    }), B.length !== L.length && console.warn("Invalid selection", g);
  }, T = (g) => {
    if (A())
      return !1;
    const { selected: b } = o.get();
    b.some(({ id: B }) => g.includes(B)) && o.set({ selected: b.filter(({ id: B }) => !g.includes(B)) });
  }, l = (g) => {
    s = g, m(o.get().selected.map(({ id: b }) => b));
  }, v = (g) => K(g, s, n);
  return e.observe(
    ({ changes: g }) => T((g.deleted || []).map((b) => b.id))
  ), {
    get event() {
      const g = o.get();
      return g ? g.event : null;
    },
    get selected() {
      const g = o.get();
      return g ? [...g.selected] : null;
    },
    get userSelectAction() {
      return s;
    },
    clear: f,
    evalSelectAction: v,
    isEmpty: A,
    isSelected: w,
    setSelected: m,
    setUserSelectAction: l,
    subscribe: o.subscribe.bind(o),
    userSelect: C
  };
}, K = (e, t, n) => {
  const o = n ? n.serialize(e) : e;
  return typeof t == "function" ? t(o) : t || "EDIT";
}, R = [];
for (let e = 0; e < 256; ++e)
  R.push((e + 256).toString(16).slice(1));
function ee(e, t = 0) {
  return (R[e[t + 0]] + R[e[t + 1]] + R[e[t + 2]] + R[e[t + 3]] + "-" + R[e[t + 4]] + R[e[t + 5]] + "-" + R[e[t + 6]] + R[e[t + 7]] + "-" + R[e[t + 8]] + R[e[t + 9]] + "-" + R[e[t + 10]] + R[e[t + 11]] + R[e[t + 12]] + R[e[t + 13]] + R[e[t + 14]] + R[e[t + 15]]).toLowerCase();
}
const te = new Uint8Array(16);
function ne() {
  return crypto.getRandomValues(te);
}
function H(e, t, n) {
  return crypto.randomUUID ? crypto.randomUUID() : oe(e);
}
function oe(e, t, n) {
  var s;
  e = e || {};
  const o = e.random ?? ((s = e.rng) == null ? void 0 : s.call(e)) ?? ne();
  if (o.length < 16)
    throw new Error("Random bytes length must be >= 16");
  return o[6] = o[6] & 15 | 64, o[8] = o[8] & 63 | 128, ee(o);
}
const De = (e) => {
  const { creator: t, updatedBy: n } = e.target, o = e.bodies.reduce((s, f) => [...s, f.creator, f.updatedBy].filter(Boolean), []);
  return [
    t,
    n,
    ...o
  ].filter((s) => s);
}, V = (e) => {
  const t = (n) => {
    const o = { ...n };
    return n.created && typeof n.created == "string" && (o.created = new Date(n.created)), n.updated && typeof n.updated == "string" && (o.updated = new Date(n.updated)), o;
  };
  return {
    ...e,
    bodies: (e.bodies || []).map(t),
    target: t(e.target)
  };
}, Re = (e, t, n, o) => ({
  id: H(),
  annotation: typeof e == "string" ? e : e.id,
  created: n || /* @__PURE__ */ new Date(),
  creator: o,
  ...t
}), se = (e, t) => {
  const n = new Set(e.bodies.map((o) => o.id));
  return t.bodies.filter((o) => !n.has(o.id));
}, ie = (e, t) => {
  const n = new Set(t.bodies.map((o) => o.id));
  return e.bodies.filter((o) => !n.has(o.id));
}, de = (e, t) => t.bodies.map((n) => {
  const o = e.bodies.find((s) => s.id === n.id);
  return { newBody: n, oldBody: o && !$(o, n) ? o : void 0 };
}).filter(({ oldBody: n }) => n).map(({ oldBody: n, newBody: o }) => ({ oldBody: n, newBody: o })), re = (e, t) => !$(e.target, t.target), W = (e, t) => {
  const n = se(e, t), o = ie(e, t), s = de(e, t);
  return {
    oldValue: e,
    newValue: t,
    bodiesCreated: n.length > 0 ? n : void 0,
    bodiesDeleted: o.length > 0 ? o : void 0,
    bodiesUpdated: s.length > 0 ? s : void 0,
    targetUpdated: re(e, t) ? { oldTarget: e.target, newTarget: t.target } : void 0
  };
};
var ae = /* @__PURE__ */ ((e) => (e.BODY_ONLY = "BODY_ONLY", e.TARGET_ONLY = "TARGET_ONLY", e))(ae || {}), O = /* @__PURE__ */ ((e) => (e.LOCAL = "LOCAL", e.REMOTE = "REMOTE", e.SILENT = "SILENT", e))(O || {});
const ce = (e, t) => {
  var f, A;
  const { changes: n, origin: o } = t;
  if (!(e.options.origin ? e.options.origin === o : o !== "SILENT"))
    return !1;
  if (e.options.ignore) {
    const { ignore: w } = e.options, C = (T) => T && T.length > 0;
    if (!(C(n.created) || C(n.deleted))) {
      const T = (f = n.updated) == null ? void 0 : f.some((v) => C(v.bodiesCreated) || C(v.bodiesDeleted) || C(v.bodiesUpdated)), l = (A = n.updated) == null ? void 0 : A.some((v) => v.targetUpdated);
      if (w === "BODY_ONLY" && T && !l || w === "TARGET_ONLY" && l && !T)
        return !1;
    }
  }
  if (e.options.annotations) {
    const w = /* @__PURE__ */ new Set([
      ...(n.created || []).map((m) => m.id),
      ...(n.deleted || []).map((m) => m.id),
      ...(n.updated || []).map(({ oldValue: m }) => m.id)
    ]);
    return !!(Array.isArray(e.options.annotations) ? e.options.annotations : [e.options.annotations]).find((m) => w.has(m));
  } else
    return !0;
}, le = (e, t) => {
  const n = new Set((e.created || []).map((l) => l.id)), o = new Set((e.updated || []).map(({ newValue: l }) => l.id)), s = new Set((t.created || []).map((l) => l.id)), f = new Set((t.deleted || []).map((l) => l.id)), A = new Set((t.updated || []).map(({ oldValue: l }) => l.id)), w = new Set((t.updated || []).filter(({ oldValue: l }) => n.has(l.id) || o.has(l.id)).map(({ oldValue: l }) => l.id)), C = [
    ...(e.created || []).filter((l) => !f.has(l.id)).map((l) => A.has(l.id) ? t.updated.find(({ oldValue: v }) => v.id === l.id).newValue : l),
    ...t.created || []
  ], m = [
    ...(e.deleted || []).filter((l) => !s.has(l.id)),
    ...(t.deleted || []).filter((l) => !n.has(l.id))
  ], T = [
    ...(e.updated || []).filter(({ newValue: l }) => !f.has(l.id)).map((l) => {
      const { oldValue: v, newValue: g } = l;
      if (A.has(g.id)) {
        const b = t.updated.find((L) => L.oldValue.id === g.id).newValue;
        return W(v, b);
      } else
        return l;
    }),
    ...(t.updated || []).filter(({ oldValue: l }) => !w.has(l.id))
  ];
  return { created: C, deleted: m, updated: T };
}, Y = (e) => {
  const t = e.id === void 0 ? H() : e.id;
  return {
    ...e,
    id: t,
    bodies: e.bodies === void 0 ? [] : e.bodies.map((n) => ({
      ...n,
      annotation: t
    })),
    target: {
      ...e.target,
      annotation: t
    }
  };
}, ue = (e) => e.id !== void 0, xe = () => {
  const e = /* @__PURE__ */ new Map(), t = /* @__PURE__ */ new Map(), n = [], o = (d, r = {}) => {
    n.push({ onChange: d, options: r });
  }, s = (d) => {
    const r = n.findIndex((c) => c.onChange == d);
    r > -1 && n.splice(r, 1);
  }, f = (d, r) => {
    const c = {
      origin: d,
      changes: {
        created: r.created || [],
        updated: r.updated || [],
        deleted: r.deleted || []
      },
      state: [...e.values()]
    };
    n.forEach((p) => {
      ce(p, c) && p.onChange(c);
    });
  }, A = (d, r = O.LOCAL) => {
    if (d.id && e.get(d.id))
      throw Error(`Cannot add annotation ${d.id} - exists already`);
    {
      const p = Y(d);
      e.set(p.id, p), p.bodies.forEach((U) => t.set(U.id, p.id)), f(r, { created: [p] });
    }
  }, w = (d, r) => {
    const c = Y(typeof d == "string" ? r : d), p = typeof d == "string" ? d : d.id, U = p && e.get(p);
    if (U) {
      const y = W(U, c);
      return p === c.id ? e.set(p, c) : (e.delete(p), e.set(c.id, c)), U.bodies.forEach((D) => t.delete(D.id)), c.bodies.forEach((D) => t.set(D.id, c.id)), y;
    } else
      console.warn(`Cannot update annotation ${p} - does not exist`);
  }, C = (d, r = O.LOCAL, c = O.LOCAL) => {
    const p = ue(r) ? c : r, U = w(d, r);
    U && f(p, { updated: [U] });
  }, m = (d, r = O.LOCAL) => {
    !!e.get(d.id) ? C(d, r) : A(d, r);
  }, T = (d, r = O.LOCAL) => {
    const c = d.reduce((p, U) => {
      const y = w(U);
      return y ? [...p, y] : p;
    }, []);
    c.length > 0 && f(r, { updated: c });
  }, l = (d, r = O.LOCAL) => {
    const c = d.map(Y), { toAdd: p, toUpdate: U } = c.reduce((D, I) => e.get(I.id) ? { ...D, toUpdate: [...D.toUpdate, I] } : { ...D, toAdd: [...D.toAdd, I] }, { toAdd: [], toUpdate: [] }), y = U.map((D) => w(D, r)).filter(Boolean);
    p.forEach((D) => {
      e.set(D.id, D), D.bodies.forEach((I) => t.set(I.id, D.id));
    }), f(r, { created: p, updated: y });
  }, v = (d, r = O.LOCAL) => {
    const c = e.get(d.annotation);
    if (c) {
      const p = {
        ...c,
        bodies: [...c.bodies, d]
      };
      e.set(c.id, p), t.set(d.id, p.id), f(r, { updated: [{
        oldValue: c,
        newValue: p,
        bodiesCreated: [d]
      }] });
    } else
      console.warn(`Attempt to add body to missing annotation: ${d.annotation}`);
  }, g = () => [...e.values()], b = (d = O.LOCAL) => {
    const r = [...e.values()];
    e.clear(), t.clear(), f(d, { deleted: r });
  }, L = (d, r = !0, c = O.LOCAL) => {
    const p = d.map(Y);
    if (r) {
      const U = [...e.values()];
      e.clear(), t.clear(), p.forEach((y) => {
        e.set(y.id, y), y.bodies.forEach((D) => t.set(D.id, y.id));
      }), f(c, { created: p, deleted: U });
    } else {
      const U = d.reduce((y, D) => {
        const I = D.id && e.get(D.id);
        return I ? [...y, I] : y;
      }, []);
      if (U.length > 0)
        throw Error(`Bulk insert would overwrite the following annotations: ${U.map((y) => y.id).join(", ")}`);
      p.forEach((y) => {
        e.set(y.id, y), y.bodies.forEach((D) => t.set(D.id, y.id));
      }), f(c, { created: p });
    }
  }, B = (d) => {
    const r = typeof d == "string" ? d : d.id, c = e.get(r);
    if (c)
      return e.delete(r), c.bodies.forEach((p) => t.delete(p.id)), c;
    console.warn(`Attempt to delete missing annotation: ${r}`);
  }, i = (d, r = O.LOCAL) => {
    const c = B(d);
    c && f(r, { deleted: [c] });
  }, u = (d, r = O.LOCAL) => {
    const c = d.reduce((p, U) => {
      const y = B(U);
      return y ? [...p, y] : p;
    }, []);
    c.length > 0 && f(r, { deleted: c });
  }, h = (d) => {
    const r = e.get(d.annotation);
    if (r) {
      const c = r.bodies.find((p) => p.id === d.id);
      if (c) {
        t.delete(c.id);
        const p = {
          ...r,
          bodies: r.bodies.filter((y) => y.id !== d.id)
        };
        return e.set(r.id, p), {
          oldValue: r,
          newValue: p,
          bodiesDeleted: [c]
        };
      } else
        console.warn(`Attempt to delete missing body ${d.id} from annotation ${d.annotation}`);
    } else
      console.warn(`Attempt to delete body from missing annotation ${d.annotation}`);
  }, E = (d, r = O.LOCAL) => {
    const c = h(d);
    c && f(r, { updated: [c] });
  }, a = (d, r = O.LOCAL) => {
    const c = d.map((p) => h(p)).filter(Boolean);
    c.length > 0 && f(r, { updated: c });
  }, S = (d) => {
    const r = e.get(d);
    return r ? { ...r } : void 0;
  }, x = (d) => {
    const r = t.get(d);
    if (r) {
      const p = S(r).bodies.find((U) => U.id === d);
      if (p)
        return p;
      console.error(`Store integrity error: body ${d} in index, but not in annotation`);
    } else
      console.warn(`Attempt to retrieve missing body: ${d}`);
  }, k = (d, r) => {
    if (d.annotation !== r.annotation)
      throw "Annotation integrity violation: annotation ID must be the same when updating bodies";
    const c = e.get(d.annotation);
    if (c) {
      const p = c.bodies.find((y) => y.id === d.id), U = {
        ...c,
        bodies: c.bodies.map((y) => y.id === p.id ? r : y)
      };
      return e.set(c.id, U), p.id !== r.id && (t.delete(p.id), t.set(r.id, U.id)), {
        oldValue: c,
        newValue: U,
        bodiesUpdated: [{ oldBody: p, newBody: r }]
      };
    } else
      console.warn(`Attempt to add body to missing annotation ${d.annotation}`);
  }, Q = (d, r, c = O.LOCAL) => {
    const p = k(d, r);
    p && f(c, { updated: [p] });
  }, J = (d, r = O.LOCAL) => {
    const c = d.map((p) => k({ id: p.id, annotation: p.annotation }, p)).filter(Boolean);
    f(r, { updated: c });
  }, j = (d) => {
    const r = e.get(d.annotation);
    if (r) {
      const c = {
        ...r,
        target: {
          ...r.target,
          ...d
        }
      };
      return e.set(r.id, c), {
        oldValue: r,
        newValue: c,
        targetUpdated: {
          oldTarget: r.target,
          newTarget: d
        }
      };
    } else
      console.warn(`Attempt to update target on missing annotation: ${d.annotation}`);
  };
  return {
    addAnnotation: A,
    addBody: v,
    all: g,
    bulkAddAnnotations: L,
    bulkDeleteAnnotations: u,
    bulkDeleteBodies: a,
    bulkUpdateAnnotations: T,
    bulkUpdateBodies: J,
    bulkUpdateTargets: (d, r = O.LOCAL) => {
      const c = d.map((p) => j(p)).filter(Boolean);
      c.length > 0 && f(r, { updated: c });
    },
    bulkUpsertAnnotations: l,
    clear: b,
    deleteAnnotation: i,
    deleteBody: E,
    getAnnotation: S,
    getBody: x,
    observe: o,
    unobserve: s,
    updateAnnotation: C,
    updateBody: Q,
    updateTarget: (d, r = O.LOCAL) => {
      const c = j(d);
      c && f(r, { updated: [c] });
    },
    upsertAnnotation: m
  };
};
let q = () => ({
  emit(e, ...t) {
    for (let n = this.events[e] || [], o = 0, s = n.length; o < s; o++)
      n[o](...t);
  },
  events: {},
  on(e, t) {
    var n;
    return ((n = this.events)[e] || (n[e] = [])).push(t), () => {
      var o;
      this.events[e] = (o = this.events[e]) == null ? void 0 : o.filter((s) => t !== s);
    };
  }
});
const fe = 250, Ne = (e, t) => {
  const n = q(), o = (t == null ? void 0 : t.changes) || [];
  let s = t ? t.pointer : -1, f = !1, A = 0;
  const w = (a) => {
    if (!f) {
      const { changes: S } = a, x = performance.now();
      if (x - A > fe)
        o.splice(s + 1), o.push(S), s = o.length - 1;
      else {
        const k = o.length - 1;
        o[k] = le(o[k], S);
      }
      A = x;
    }
    f = !1;
  };
  e.observe(w, { origin: O.LOCAL });
  const C = (a) => a && a.length > 0 && e.bulkDeleteAnnotations(a), m = (a) => a && a.length > 0 && e.bulkAddAnnotations(a, !1), T = (a) => a && a.length > 0 && e.bulkUpdateAnnotations(a.map(({ oldValue: S }) => S)), l = (a) => a && a.length > 0 && e.bulkUpdateAnnotations(a.map(({ newValue: S }) => S)), v = (a) => a && a.length > 0 && e.bulkAddAnnotations(a, !1), g = (a) => a && a.length > 0 && e.bulkDeleteAnnotations(a);
  return {
    canRedo: () => o.length - 1 > s,
    canUndo: () => s > -1,
    destroy: () => e.unobserve(w),
    getHistory: () => ({ changes: [...o], pointer: s }),
    on: (a, S) => n.on(a, S),
    redo: () => {
      if (o.length - 1 > s) {
        f = !0;
        const { created: a, updated: S, deleted: x } = o[s + 1];
        m(a), l(S), g(x), n.emit("redo", o[s + 1]), s += 1;
      }
    },
    undo: () => {
      if (s > -1) {
        f = !0;
        const { created: a, updated: S, deleted: x } = o[s];
        C(a), T(S), v(x), n.emit("undo", o[s]), s -= 1;
      }
    }
  };
}, ke = () => {
  const e = /* @__PURE__ */ M([]);
  return {
    subscribe: e.subscribe.bind(e),
    set: e.set.bind(e)
  };
}, Ie = (e, t, n, o) => {
  const { hover: s, selection: f, store: A, viewport: w } = e, C = /* @__PURE__ */ new Map();
  let m = [], T, l;
  const v = (i, u) => {
    C.has(i) ? C.get(i).push(u) : C.set(i, [u]);
  }, g = (i, u) => {
    const h = C.get(i);
    if (h) {
      const E = h.indexOf(u);
      E !== -1 && h.splice(E, 1);
    }
  }, b = (i, u, h) => {
    C.has(i) && setTimeout(() => {
      C.get(i).forEach((E) => {
        if (n) {
          const a = Array.isArray(u) ? u.map((x) => n.serialize(x)) : n.serialize(u), S = h ? h instanceof PointerEvent ? h : n.serialize(h) : void 0;
          E(a, S);
        } else
          E(u, h);
      });
    }, 1);
  }, L = () => {
    const { selected: i } = f, u = (i || []).map(({ id: h }) => A.getAnnotation(h));
    u.forEach((h) => {
      const E = m.find((a) => a.id === h.id);
      (!E || !$(E, h)) && b("updateAnnotation", h, E);
    }), m = m.map((h) => {
      const E = u.find(({ id: a }) => a === h.id);
      return E || h;
    });
  };
  f.subscribe(({ selected: i }) => {
    if (!(m.length === 0 && i.length === 0)) {
      if (m.length === 0 && i.length > 0)
        m = i.map(({ id: u }) => A.getAnnotation(u));
      else if (m.length > 0 && i.length === 0)
        m.forEach((u) => {
          const h = A.getAnnotation(u.id);
          h && !$(h, u) && b("updateAnnotation", h, u);
        }), m = [];
      else {
        const u = new Set(m.map((a) => a.id)), h = new Set(i.map(({ id: a }) => a));
        m.filter((a) => !h.has(a.id)).forEach((a) => {
          const S = A.getAnnotation(a.id);
          S && !$(S, a) && b("updateAnnotation", S, a);
        }), m = [
          // Remove annotations that were deselected
          ...m.filter((a) => h.has(a.id)),
          // Add editable annotations that were selected
          ...i.filter(({ id: a }) => !u.has(a)).map(({ id: a }) => A.getAnnotation(a))
        ];
      }
      b("selectionChanged", m);
    }
  }), s.subscribe((i) => {
    !T && i ? b("mouseEnterAnnotation", A.getAnnotation(i)) : T && !i ? b("mouseLeaveAnnotation", A.getAnnotation(T)) : T && i && (b("mouseLeaveAnnotation", A.getAnnotation(T)), b("mouseEnterAnnotation", A.getAnnotation(i))), T = i;
  }), w == null || w.subscribe((i) => b("viewportIntersect", i.map((u) => A.getAnnotation(u)))), A.observe((i) => {
    o && (l && clearTimeout(l), l = setTimeout(L, 1e3));
    const { created: u, deleted: h } = i.changes;
    (u || []).forEach((a) => b("createAnnotation", a)), (h || []).forEach((a) => b("deleteAnnotation", a)), (i.changes.updated || []).filter((a) => [
      ...a.bodiesCreated || [],
      ...a.bodiesDeleted || [],
      ...a.bodiesUpdated || []
    ].length > 0).forEach(({ oldValue: a, newValue: S }) => {
      const x = m.find((k) => k.id === a.id) || a;
      m = m.map((k) => k.id === a.id ? S : k), b("updateAnnotation", S, x);
    });
  }, { origin: O.LOCAL }), A.observe((i) => {
    if (m) {
      const u = new Set(m.map((E) => E.id)), h = (i.changes.updated || []).filter(({ newValue: E }) => u.has(E.id)).map(({ newValue: E }) => E);
      h.length > 0 && (m = m.map((E) => {
        const a = h.find((S) => S.id === E.id);
        return a || E;
      }));
    }
  }, { origin: O.REMOTE });
  const B = (i) => (u) => {
    const { updated: h } = u;
    i ? (h || []).forEach((E) => b("updateAnnotation", E.oldValue, E.newValue)) : (h || []).forEach((E) => b("updateAnnotation", E.newValue, E.oldValue));
  };
  return t.on("undo", B(!0)), t.on("redo", B(!1)), { on: v, off: g, emit: b };
}, _e = (e) => (t) => t.map((n) => e.serialize(n)), pe = (e) => (t) => t.reduce((n, o) => {
  const { parsed: s, error: f } = e.parse(o);
  return f ? {
    parsed: n.parsed,
    failed: [...n.failed, o]
  } : s ? {
    parsed: [...n.parsed, s],
    failed: n.failed
  } : {
    ...n
  };
}, { parsed: [], failed: [] }), $e = (e, t, n) => {
  const { store: o, selection: s } = e, f = (i) => {
    if (n) {
      const { parsed: u, error: h } = n.parse(i);
      u ? o.addAnnotation(u, O.REMOTE) : console.error(h);
    } else
      o.addAnnotation(V(i), O.REMOTE);
  }, A = () => s.clear(), w = () => o.clear(), C = (i) => {
    const u = o.getAnnotation(i);
    return n && u ? n.serialize(u) : u;
  }, m = () => n ? o.all().map(n.serialize) : o.all(), T = () => {
    var h;
    const u = (((h = s.selected) == null ? void 0 : h.map((E) => E.id)) || []).map((E) => o.getAnnotation(E)).filter(Boolean);
    return n ? u.map(n.serialize) : u;
  }, l = (i, u = !0) => fetch(i).then((h) => h.json()).then((h) => (g(h, u), h)), v = (i) => {
    if (typeof i == "string") {
      const u = o.getAnnotation(i);
      if (o.deleteAnnotation(i), u)
        return n ? n.serialize(u) : u;
    } else {
      const u = n ? n.parse(i).parsed : i;
      if (u)
        return o.deleteAnnotation(u), i;
    }
  }, g = (i, u = !0) => {
    if (n) {
      const h = n.parseAll || pe(n), { parsed: E, failed: a } = h(i);
      a.length > 0 && console.warn(`Discarded ${a.length} invalid annotations`, a), o.bulkAddAnnotations(E, u, O.REMOTE);
    } else
      o.bulkAddAnnotations(i.map(V), u, O.REMOTE);
  }, b = (i, u) => {
    i ? s.setSelected(i, u) : s.clear();
  }, L = (i) => {
    s.setUserSelectAction(i);
  }, B = (i) => {
    if (n) {
      const u = n.parse(i).parsed, h = n.serialize(o.getAnnotation(u.id));
      return o.updateAnnotation(u), h;
    } else {
      const u = o.getAnnotation(i.id);
      return o.updateAnnotation(V(i)), u;
    }
  };
  return {
    addAnnotation: f,
    cancelSelected: A,
    canRedo: t.canRedo,
    canUndo: t.canUndo,
    clearAnnotations: w,
    getAnnotationById: C,
    getAnnotations: m,
    getHistory: t.getHistory,
    getSelected: T,
    loadAnnotations: l,
    redo: t.redo,
    removeAnnotation: v,
    setAnnotations: g,
    setSelected: b,
    setUserSelectAction: L,
    undo: t.undo,
    updateAnnotation: B
  };
}, Ye = (e, t, n) => typeof t == "function" ? t(e, n) : t, ze = (e, t) => typeof e != "function" && typeof t != "function" ? {
  ...e || {},
  ...t || {}
} : (n, o) => {
  const s = typeof e == "function" ? e(n, o) : e, f = typeof t == "function" ? t(n, o) : t;
  return {
    ...s || {},
    ...f || {}
  };
};
let ge = "useandom-26T198340PX75pxJACKVERYMINDBUSHWOLF_GQZbfghjklqvwyzrict", he = (e) => crypto.getRandomValues(new Uint8Array(e)), Ae = (e, t, n) => {
  let o = 256 - 256 % e.length;
  if (o === 256) {
    let f = e.length - 1;
    return (A = t) => {
      if (!A) return "";
      let w = "";
      for (; ; ) {
        let C = n(A), m = A;
        for (; m--; )
          if (w += e[C[m] & f], w.length >= A) return w;
      }
    };
  }
  let s = Math.ceil(1.6 * 256 * t / o);
  return (f = t) => {
    if (!f) return "";
    let A = "";
    for (; ; ) {
      let w = n(s), C = s;
      for (; C--; )
        if (w[C] < o && (A += e[w[C] % e.length], A.length >= f))
          return A;
    }
  };
}, me = (e, t = 21) => Ae(e, t | 0, he), be = (e = 21) => {
  let t = "", n = crypto.getRandomValues(new Uint8Array(e |= 0));
  for (; e--; )
    t += ge[n[e] & 63];
  return t;
};
const Pe = () => ({ isGuest: !0, id: me("1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_", 20)() }), ve = (e) => {
  const t = JSON.stringify(e);
  let n = 0;
  for (let o = 0, s = t.length; o < s; o++) {
    let f = t.charCodeAt(o);
    n = (n << 5) - n + f, n |= 0;
  }
  return `${n}`;
}, Ee = (e) => e ? typeof e == "object" ? { ...e } : e : void 0, Ve = (e, t) => (Array.isArray(e) ? e : [e]).map((n) => {
  const { id: o, type: s, purpose: f, value: A, created: w, modified: C, creator: m, ...T } = n;
  return {
    id: o || `temp-${ve(n)}`,
    annotation: t,
    type: s,
    purpose: f,
    value: A,
    creator: Ee(m),
    created: w ? new Date(w) : void 0,
    updated: C ? new Date(C) : void 0,
    ...T
  };
}), Me = (e) => e.map((t) => {
  var w;
  const { annotation: n, created: o, updated: s, ...f } = t, A = {
    ...f,
    created: o == null ? void 0 : o.toISOString(),
    modified: s == null ? void 0 : s.toISOString()
  };
  return (w = A.id) != null && w.startsWith("temp-") && delete A.id, A;
}), je = (e) => "@context" in e && "creator" in e && "body" in e, we = [
  "#ff7c00",
  // orange
  "#1ac938",
  // green
  "#e8000b",
  // red
  "#8b2be2",
  // purple
  "#9f4800",
  // brown
  "#f14cc1",
  // pink
  "#ffc400",
  // khaki
  "#00d7ff",
  // cyan
  "#023eff"
  // blue
], Ce = () => {
  const e = [...we];
  return { assignRandomColor: () => {
    const o = Math.floor(Math.random() * e.length), s = e[o];
    return e.splice(o, 1), s;
  }, releaseColor: (o) => e.push(o) };
}, Le = () => {
  const e = Ce();
  return { addUser: (o, s) => {
    const f = e.assignRandomColor();
    return {
      label: s.name || s.id,
      avatar: s.avatar,
      color: f
    };
  }, removeUser: (o) => e.releaseColor(o.appearance.color) };
}, ye = (e, t) => e.every((n) => e.includes(n)) && t.every((n) => e.includes(n)), Ge = be(), He = (e = Le()) => {
  const t = q(), n = /* @__PURE__ */ new Map(), o = /* @__PURE__ */ new Map(), s = (l, v) => {
    if (n.has(l)) {
      console.warn("Attempt to add user that is already present", l, v);
      return;
    }
    const g = e.addUser(l, v);
    n.set(l, {
      ...v,
      presenceKey: l,
      appearance: g
    });
  }, f = (l) => {
    const v = n.get(l);
    if (!v) {
      console.warn("Attempt to remove user that is not present", l);
      return;
    }
    e.removeUser(v), n.delete(l);
  }, A = (l) => {
    const v = new Set(l.map((L) => L.presenceKey)), g = l.filter(({ presenceKey: L }) => !n.has(L)), b = Array.from(n.values()).filter((L) => !v.has(L.presenceKey));
    g.forEach(({ presenceKey: L, user: B }) => s(L, B)), b.forEach((L) => {
      const { presenceKey: B } = L;
      o.has(B) && t.emit("selectionChange", L, null), f(B);
    }), (g.length > 0 || b.length > 0) && t.emit("presence", m());
  }, w = (l, v) => {
    const g = n.get(l);
    if (!g) {
      console.warn("Activity notification from user that is not present");
      return;
    }
    const b = o.get(l);
    (!b || !ye(b, v)) && (o.set(l, v), t.emit("selectionChange", g, v));
  }, C = (l, v) => {
    const g = n.get(l);
    if (!g) {
      console.warn("Selection change for user that is not present", l);
      return;
    }
    v ? o.set(l, v) : o.delete(l), t.emit("selectionChange", g, v);
  }, m = () => [...Array.from(n.values())];
  return {
    getPresentUsers: m,
    notifyActivity: w,
    on: (l, v) => t.on(l, v),
    syncUsers: A,
    updateSelection: C
  };
};
export {
  ae as Ignore,
  O as Origin,
  Ge as PRESENCE_KEY,
  F as UserSelectAction,
  ze as chainStyles,
  Ye as computeStyle,
  Pe as createAnonymousGuest,
  $e as createBaseAnnotator,
  Re as createBody,
  Le as createDefaultAppearanceProvider,
  Oe as createHoverState,
  Ie as createLifecycleObserver,
  He as createPresenceState,
  Be as createSelectionState,
  xe as createStore,
  Ne as createUndoStack,
  ke as createViewportState,
  Ce as defaultColorProvider,
  W as diffAnnotations,
  De as getContributors,
  je as isW3CAnnotation,
  le as mergeChanges,
  K as onUserSelect,
  pe as parseAll,
  Ve as parseW3CBodies,
  Ee as parseW3CUser,
  V as reviveDates,
  _e as serializeAll,
  Me as serializeW3CBodies,
  ce as shouldNotify
};
